src/trace.h

```diff
@@ -38,14 +41,6 @@ GIT_INLINE(void) git_trace__write_fmt(
 	git_buf_dispose(&message);
 }
 
-#define git_trace_level()		(git_trace__data.level)
-#define git_trace(l, ...)		{ \
-									if (git_trace__data.level >= l && \
-										git_trace__data.callback != NULL) { \
-										git_trace__write_fmt(l, __VA_ARGS__); \
-									} \
-								}
```

Comment:
The ironic thing is: the only thing that understandably keeps you from accepting this is also the only real issue we have according to your statement, as it's the only place where we expose a non-C89 API :)

I'm trying to think of any way to implement this in a non-variadic way, but I cannot really think of any. We could at least force-inline it, but that's not a proper solution, either:

```
diff --git a/src/trace.h b/src/trace.h
index ed0fa0949..8c3afa7f3 100644
--- a/src/trace.h
+++ b/src/trace.h
@@ -23,27 +23,37 @@ extern struct git_trace_data git_trace__data;
 
 #define git_trace_level() (git_trace__data.level)
 
-GIT_INLINE(void) git_trace(git_trace_level_t level, const char *fmt, ...)
+GIT_INLINE(void) git_trace__write_fmt(git_trace_level_t level, const char *fmt, va_list ap);
 {
 	git_trace_cb callback = git_trace__data.callback;
 	git_buf message = GIT_BUF_INIT;
-	va_list ap;
 
-	if (git_trace__data.level < level || !git_trace__data.callback)
-		return;
-
-	va_start(ap, fmt);
 	git_buf_vprintf(&message, fmt, ap);
-	va_end(ap);
-
 	callback(level, git_buf_cstr(&message));
-
 	git_buf_dispose(&message);
 }
 
+#if defined(_MSC_VER)
+# define GIT_ALWAYS_INLINE(type) static __forceinline type
+#elif defined(__GNUC__)
+# define GIT_ALWAYS_INLINE(type) static inline __attribute__((__always_inline__)) type
 #else
+# define GIT_ALWAYS_INLINE(type) static type
+#endif
 
-GIT_INLINE(void) git_trace__null(
+GIT_ALWAYS_INLINE(void) git_trace(git_trace_level_t level, const char *fmt, ...)
+{
+	if (git_trace__data.level >= level && git_trace__data.callback != NULL) {
+		va_list ap;
+		va_start(ap, fmt);
+		git_trace__write_fmt(l, fmt, ap);
+		va_end(ap);
+	}
+}
+
+#else
+
+GIT_ALWAYSINLINE(void) git_trace__null(
 	git_trace_level_t level,
 	const char *fmt, ...)
 {
```

Reply:
> The ironic thing is: the only thing that understandably keeps you from accepting this is also the only real issue we have according to your statement, as it's the only place where we expose a non-C89 API :)

It's not exposed to end-users, though, `git_trace` is only ever used internally.  (The tracing setup is exposed, but not variadic.)

Reply:
Sorry if I'm missing your point. `git_trace` is exposed via "include/git2/trace.h", so I'd very much argue that it is exposed to end-users.

Reply:
But not the variadic bits - only `git_trace_set`, the levels enum and the callback definition is public.

Reply:
Gotcha, somehow I've missed that. So yeah, as I don't have any fix available, should I just drop the tracing part of this PR for now and at least get the other parts in?

Reply:
Oh jeez I missed this question.  I'm +1 on doing this except for the tracing bits.  As I mentioned, I'd like to have a play with the perf on changes there, and it's definitely a high priority for me.  Unfortunately, that means ...  a little while still?  :/